rfc/system/0000-caching-of-offloaded-objects.md

```diff
@@ -0,0 +1,161 @@
+
+### Cache key calculation
+
+As mentioned above, during the regular cache key calculation, in both the local and the remote cases, flyte takes the input literal map as part of the key. Going forward we will allow for specific literals in that map to use the overridden hash as the representation of that literal.
+
+### Other clients
+
+Although nothing prevents the adoption of this feature in other clients, flytekit-java and other clients are outside of the scope of this RFC.
+
+## 4 Metrics & Dashboards
+
+N/A?
+
+## 5 Drawbacks
+
+Although this feature does *not* impact the other aspects of how the cache works, for example,  changes to the version or the signature of the task invalidate the cache entries, it still might cause confusion since from the perspective of a task execution we will not be able to tell if an object will have its hash overridden. 
+
+## 6 Alternatives
+
+A few options were discussed in https://github.com/flyteorg/flyte/issues/1581:
+
+**Expose a `hash` method in Type Transformers and annotate the task in case the type transformer does not contain a hash function:** The major drawbacks of this alternative are two-fold: 
+    (1) There's no way to opt-out, i.e. all objects produced by that Type Transformer will be cached.
+    (2) The UX is a bit clunky in the case of user-defined hash functions for two reasons:
+        - the return type does not contain any indication that the object is being cached, i.e. we'd have to look in the parameters set in the @task decorator
+
+**Offload the hash calculation to an external system like s3:** The idea would be to rely on etags produced by the act of storing an object in s3. This turns out to be a chicken-and-egg problem as we need to produce etags in order to compare if that tag was already produced. Also, even if we were able to overcome that hurdle, handing this computation to a blackbox system like s3 risks making cache entries unrecoverable, since there is no guarantee that the algorithm s3 uses to compute etags will not change.
+
+**Define the new hash field in the Scalar instead of the Literal**: One could argue that the hash could be "tucked away" in the [Scalar](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/literals.proto#L55-L65) instead. At this time I couldn't find a great argument for going that route, although most of the reasoning around the use of the hash during cache key calculation still applies (adjusting the access to the hash field).
+
+## 7 Potential Impact and Dependencies
+
+It might be useful to cap the size of hashes produced by client code so as to limit the size of the cache keys. Performing this capping should be done in an uniform manner, which implies that it should happen in the backend.
+
+## 8 Unresolved questions
+
+How does this affect @dynamic tasks, subworkflows, and launchplans?
+
+What's the observability provided by data catalog?
```

Comment:
Not sure what does it mean to have an input with that annotation? do you mean flyte/flytekit would log that "task author expected hash and you provided non..." ?

Reply:
> do you mean flyte/flytekit would log that "task author expected hash and you provided non..." ?

Exactly.

Reply:
If we leverage tags as we seem to be [leaning towards](https://github.com/flyteorg/flyte/pull/1893#discussion_r763364855), maybe this is a moot point then? because by reference lookup will continue to work as expected? I think if we surface the information to the user (that we used did a lookup by reference not by hash) maybe that is enough? not as a warning or anything... just stating a fact...
We will have to thoroughly document all of this... it's a good pattern if we can make sure users understand what's going on...

Reply:
I still think that it's worth showing to the user whenever we are able that this caching-by-value semantics is used. 

One idea we talked about is to catch cases as early as during registration phase and at least inform the user that their intent might not match the code. I updated the RFC with two examples of such ideas.